### DISCUSSION TITLE
Update State Programmatically

### DISCUSSION THREAD

**Discussion by frederickparsons-tandemai** [molstar/molstar#1108] [General]
Hi molstar community!

I really like what is being done with the Molviewspec and with the Molstar State. It really opens a declarative way of interacting with the viewer!

Let's suppose I wanted to perform all viewer interactions through the molstar state object. I would not be using the wrapper UI - just the molstar viewer itself. I would programmatically modify the state (add loci, change representation etc...) imperatively through user actions.

Is it possible to modify the state object directly and have the viewer react instead of using the imperative functions. In other words, my wrapper application would know how to modify the State and then would call something like `molstar.setState(newState)`. Is this possible?

For example, instead of calling `plugin.managers.interactivity.lociHighlights.highlightOnly({ loci })` I'd modify the state object directly.

**Comment by dsehnal** [comment]
Hi Fred,

the native state manager has the option to serialize current selection (<strike>albeit I just noticed a bug where empty selection isn't correctly applied when moving from a state with existing selection; but it should be an easy fix</strike> actually, no bug, I just forgot to enable the feature for the 1st snapshot I made):

![image](https://github.com/molstar/molstar/assets/21145776/e78e78be-f4be-4b3a-98c0-a4cddeda6424)
![image](https://github.com/molstar/molstar/assets/21145776/fe994722-3a83-41e2-8960-000f0c4c9350)

Technically, same functionality could be added for the highlight state as well -- although at the end of the day, it would just end up calling `highlightOnly` for you.

In principle, MVS could also be extended in this regard, although I would have to think about this more.

**Reply by frederickparsons-tandemai** [reply]
Hi David thanks for the reply. Initially we considered using the MVS however it doesn't contain the complete state that we'd want to serialize. For example right now certain representations are not supported and we can't store the measurements. Would you recommend that we serialize both the MVS and the state to fully recover all the session details?

**Reply by dsehnal** [reply]
MVS functionality is currently quite limited. We do have plans to include things like measurements or custom primitives (drawing lines, meshes, etc.), but not specific timeline for that currently. Also, the goal for MVS is to be interpretable by viewers other than Mol*, so we want to be somewhat careful with what functionality we include.

I am hesitant to recommend you anything specific as I don't know the requirements for your use case, how extensible you want your solution to be, etc.
